### PR TITLE
TDL-19330: Add error logging

### PR DESCRIPTION
**Pull request by hpatel41**
# Description of change
TDL-19330: Add error logging
- Added custom error handling as per recharge error codes. ([Reference](https://docs.rechargepayments.com/docs/api-standards#errors))

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_recharge/client.py [hunk 1]
```diff
@@ -52,51 +56,117 @@ class RechargeForbiddenError(RechargeError):
...
+
+    error_message = response_json.get('error', response_json.get('errors'))
+
+    if not error_message:
+        error_message = ERROR_CODE_EXCEPTION_MAPPING.get(error_code, {}).get('message', 'Unknown Error')
+
+    if error_code == 401 and 'bad authentication' in error_message:
+        LOGGER.error("Your API Token has been deleted or the token is invalid.\n Please re-authenticate your connection to generate a new token and resume extraction.")
+
+    message = f'HTTP-error-code: {error_code}, Error: {error_message}'
```
Comment (umangagrawal-crest) [inline tap_recharge/client.py:166]:
```suggestion
    message = 'HTTP-error-code: %s, Error: %s' %(error_code,error_message)
```

File: tap_recharge/client.py [hunk 2]
```diff
@@ -52,51 +56,117 @@ class RechargeForbiddenError(RechargeError):
...
+    422: {
+        "exception": RechargeUnprocessableEntityError,
+        "message": "The request was understood but cannot be processed due to invalid or missing supplemental information.",
+    },
+    426: {
+        "exception": RechargeInvalidAPI,
+        "message": "The request was made using an invalid API version.",
+    },
+    429: {
+        "exception": Server429Error,
```
Comment (savan-chovatiya) [inline tap_recharge/client.py:118]: Replace `Server429Error` to `RechargeRateLimitError` for all occurences. Did we add backoff for 429 error?

Comment (harshpatel4crest) [reply tap_recharge/client.py]: Yes, there is backoff for 429 error. Updated the **Server429Error** to **RechargeRateLimitError**

File: tap_recharge/client.py [hunk 3]
```diff
@@ -44,59 +42,129 @@ class RechargeMethodNotAllowedError(RechargeError):
...
 
 def get_exception_for_error_code(error_code):
-    return ERROR_CODE_EXCEPTION_MAPPING.get(error_code, RechargeError)
+    """Function to retrieve exceptions based on error code"""
+    if error_code == 429:
+        # Delay for 5 seconds for leaky bucket rate limit algorithm
+        time.sleep(5)
+
+    exception = ERROR_CODE_EXCEPTION_MAPPING.get(error_code, {}).get('exception')
+    if not exception:
```
Comment (dbshah1212) [inline tap_recharge/client.py:140]: Add code comment for this condition.

File: tap_recharge/client.py [hunk 4]
```diff
@@ -44,59 +42,129 @@ class RechargeMethodNotAllowedError(RechargeError):
...
+        if error_code >= 500:
+            return Server5xxError
+        return RechargeError
+    return exception
 
 def raise_for_error(response):
+    """Function to raise custom error along with message based on response."""
+
+    error_code = response.status_code
```
Comment (dbshah1212) [inline tap_recharge/client.py:149]: Use get else it will break the tap in case of an empty response.

File: tap_recharge/client.py [hunk 5]
```diff
@@ -44,59 +42,129 @@ class RechargeMethodNotAllowedError(RechargeError):
...
+    except ValueError:
+        response_json = {}
+
+    error_message = response_json.get('error', response_json.get('errors'))
+
+    if not error_message:
+        error_message = ERROR_CODE_EXCEPTION_MAPPING.get(error_code, {}).get('message', 'Unknown Error')
+
+    if error_code == 401 and 'bad authentication' in error_message:
+        LOGGER.error("Your API Token has been deleted or the token is invalid.\n Please re-authenticate your connection to generate a new token and resume extraction.")
```
Comment (dbshah1212) [inline tap_recharge/client.py:162]: This must be just before the exception so move it 2 lines down.

Comment (harshpatel4crest) [reply tap_recharge/client.py]: Updated the location of the logger.

**Review by savan-chovatiya — APPROVED** [review]

**Review by dbshah1212 — APPROVED** [review]

File: tests/unittests/test_backoff.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+        self.assertEqual(mocked_request.call_count, 5)
+
+    def test_Server429Error_error_backoff__request(self, mocked_request, mocked_sleep, mocked_check_access_token):
+        """Test case to verify we backoff for 5 times for '429' error"""
+        mocked_request.side_effect = RechargeRateLimitError('ratelimit error')
+        with self.assertRaises(RechargeRateLimitError) as e:
+            response_json, _ = self.client_obj.request(self.method, self.path, self.url)
+
+        self.assertEqual(mocked_request.call_count, 5)
```
Comment (RushiT0122) [inline tests/unittests/test_backoff.py:59]: Can we parameterise these unittests just like we did for request timeout in PR#32?

**Review by RushiT0122 — CHANGES_REQUESTED** [review]
Added request change in the unittests/test_backoff.py.

File: tests/unittests/test_exception_handling.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        ['429_error', [429, RechargeRateLimitError], 'HTTP-error-code: 429, Error: The request has been rate limited.'],
+        ['500_error', [500, RechargeInternalServiceError], 'HTTP-error-code: 500, Error: The request could not be processed due to internal server error.'],
+        ['501_error', [501, RechargeUnimplementedResourceError], 'HTTP-error-code: 501, Error: The resource requested has not been implemented in the current version.'],
+        ['502_error', [502, Server5xxError], 'HTTP-error-code: 502, Error: Unknown Error'],
+        ['503_error', [503, RechargeThirdPartyServiceTimeoutError], 'HTTP-error-code: 503, Error: A third party service on which the request depends has timed out.']
+    ])
+    @mock.patch('tap_recharge.client.RechargeClient.check_access_token')
+    @mock.patch('time.sleep')
+    @mock.patch('requests.Session.request')
+    def test_custom_exception_handling(self, name, test_data, expected_data, mocked_request, mocked_sleep, mocked_check_token):
```
Comment (RushiT0122) [inline tests/unittests/test_exception_handling.py:112]: Here `name` parameter is unused in the test. Where is this getting used?

Comment (harshpatel4crest) [reply tests/unittests/test_exception_handling.py:112]: The `name` contains the name for the test case. Hence, the test cases will be named like, `test_custom_exception_handling_0_400_error`, `test_custom_exception_handling_1_401_error` and so on.

File: tests/unittests/test_exception_handling.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        ['429_error', [429, {'error': 'timeout error'}, RechargeRateLimitError], 'HTTP-error-code: 429, Error: timeout error'],
+        ['500_error', [500, {'error': 'internal server error'}, RechargeInternalServiceError], 'HTTP-error-code: 500, Error: internal server error'],
+        ['501_error', [501, {'error': 'not implemented'}, RechargeUnimplementedResourceError], 'HTTP-error-code: 501, Error: not implemented'],
+        ['502_error', [502, {'error': 'bad gateway error'}, Server5xxError], 'HTTP-error-code: 502, Error: bad gateway error'],
+        ['503_error', [503, {'error': 'third party service timed out'}, RechargeThirdPartyServiceTimeoutError], 'HTTP-error-code: 503, Error: third party service timed out']
+    ])
+    @mock.patch('tap_recharge.client.RechargeClient.check_access_token')
+    @mock.patch('time.sleep')
+    @mock.patch('requests.Session.request')
+    def test_API_exception_handling(self, name, test_data, expected_data, mocked_request, mocked_sleep, mocked_check_token):
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/unittests/test_exception_handling.py:76]: The `name` contains the name for the test case. Hence, the test cases will be named like, `test_API_exception_handling_0_400_error`, `test_API_exception_handling_1_401_error` and so on.

**Review by RushiT0122 — APPROVED** [review]

**Review by RushiT0122 — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

**Pull request merged**